### ISSUE TITLE
Bot managed to change an 'accepted' request back to 'review'

### ISSUE THREAD

**Issue by DimStar77** [openSUSE/open-build-service#5594]
This is a bit of a special corner case, that won't happen very often, but one of our bots managed to make a state change of a submitrequest from 'accepted' to 'review', which, IIRC< should not be possible.

The request in question is https://build.opensuse.org/request/show/628349

As I needed this as an urent fix in TW, I force-accepted (```osc rq accept -f``` it before the reviews all passed. At the same time, one of the bots already 'picked it up for inspection'; and set the request 'after my acceptance' into state 'Review'; I'd rather have expected the bot to receive a permission error or any such thing, that a state change from accepted to review was not possible.

(I have since declined the request, as I was not eager to find out what would happen if this request were to be accepted a 2nd time)

**Comment by adrianschroeter** [comment]
It is not clear here who wins, but one if accept call was successful it must indeed not possible to set it back.

**Comment by bgeuken** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Fix issue 5594](https://trello.com/c/YvaxFS4X/1502-fix-issue-5594)

**Comment by vpereira** [comment]
```
I, [2018-08-09T08:06:04.931951 #23593]  INFO -- : [4fdd0b68-08cb-48cf-bfca-880f612a4313] [23593:518.73] method=POST path=/request/628349 format=xml controller=RequestController action=request_command status=200 duration=1707.33 view=3.35 db=1631.11 params={"newstate"=>"accepted", "by_group"=>"legal-auto", "cmd"=>"changereviewstate"} host=172.16.42.1 time=2018-08-09 08:06:03 +0000 user=licensedigger
I, [2018-08-09T08:06:22.662232 #23638]  INFO -- : [90eb9bfd-1636-4068-91dd-0cd1dce89b0b] [23638:536.28] method=POST path=/request/628349 format=xml controller=RequestController action=request_command status=200 duration=3074.53 view=4.63 db=2844.08 params={"by_group"=>"opensuse-review-team", "cmd"=>"addreview"} host=172.16.42.1 time=2018-08-09 08:06:19 +0000 user=factory-auto
```

I don't see however it happening in a transaction. There is a transaction in ```assignreview``` but it wasn't called in the flow.. any idea @adrianschroeter ?

**Comment by adrianschroeter** [comment]
@vpereira is the transaction running from state change to actual change?

I would test this on CLI with a breakpoint just before the review state change. (start review state change, run request state change, continue review change from now and check the result)

**Comment by vpereira** [comment]
@adrianschroeter yes, everything is running inside a ```with_lock``` block. The only bit that is outside, is in the ```addreview``` command, https://github.com/vpereira/open-build-service/blob/master/src/api/app/models/bs_request.rb#L823 the check ```permission_check_addreview!``` which check again ithe request state is outside.

**Comment by adrianschroeter** [comment]
this is only about change_review_state here ... as I said, I would test on CLI with enabled SQL output (no trust in activerecord ;)

**Comment by vpereira** [comment]
I wasn't able to reproduce it @adrianschroeter 

what I did it:

1. set a ```binding.pry``` on https://github.com/vpereira/open-build-service/blob/master/src/api/app/models/bs_request.rb#L793 before the ```save```
2. run a ``` osc -A http://127.0.0.1:3000 rq accept  -m ok $REQID```
(it will stop on ```binding.pry```
3. in another session i run the ```osc review add -G security-team $REQID```

what I get: the request get accepted; and the ```osc review add ...``` fails with:

```
Server returned an error: HTTP Error 403: Forbidden
The request is not in state new or review
```
Probably in development with one server instance it will be impossible to reproduce it, I will get two instances of rails running in parallel..

**Comment by bgeuken** [comment]
The initial description doesn't sound to me like there were 2 requests send within a very short time frame. It seems that the actual problem is that we don't have a validation that permits changing the request state from 'accepted' to something else.

In the logs you can also see that there was a 18 seconds gap between the two requests.

> vpereira commented 22 hours ago
> 
 ```
I, [2018-08-09T08:06:04.931951 #23593]  INFO -- : [4fdd0b68-08cb-48cf-bfca-880f612a4313] [23593:518.73] method=POST path=/request/628349 format=xml controller=RequestController action=request_command status=200 duration=1707.33 view=3.35 db=1631.11 params={"newstate"=>"accepted", "by_group"=>"legal-auto", "cmd"=>"changereviewstate"} host=172.16.42.1 time=2018-08-09 08:06:03 +0000 user=licensedigger
I, [2018-08-09T08:06:22.662232 #23638]  INFO -- : [90eb9bfd-1636-4068-91dd-0cd1dce89b0b] [23638:536.28] method=POST path=/request/628349 format=xml controller=RequestController action=request_command status=200 duration=3074.53 view=4.63 db=2844.08 params={"by_group"=>"opensuse-review-team", "cmd"=>"addreview"} host=172.16.42.1 time=2018-08-09 08:06:19 +0000 user=factory-auto
 ```

I would write a test that verifies that changing an accepted request fails. And then ensure that we handle this properly in the controller (returning the correct error state and message).

**Comment by vpereira** [comment]
I think 18 seconds for hitting the controller, but it gets through delay job

**Comment by vpereira** [comment]
ok, I will do it @bgeuken

**Comment by vpereira** [comment]
@bgeuken  https://github.com/vpereira/open-build-service/commit/ca427954c5c9dcf2bb030bbfd93345f299721c5f is that ok?

**Comment by bgeuken** [comment]
I answer already in person, but in case anyone else is interested:smiley: Yes, that looks good. But I would also write a test that this is also not possible via `update_attributes` and alike. So that there is some basic (ActiveRecord) validation.

**Comment by adrianschroeter** [comment]
the check you wrote exists already in the api integration checks, so I am sure it can not happen usually.
This is really about the race of two running operations.

And you would need to test the review state change, that this one is not setting back the request state.

**Comment by hennevogel** [comment]
The journey is this:

`RequestController.request_command` -> `ApplicationController.dispatch_command` -> `RequestController.request_command_changestate` -> `BsRequest.change_state` -> transaction -> `BsRequest.permission_check_change_state` -> `BsRequestPermissionCheck.cmd_changestate_permissions` -> raise `PostRequestNoPermission` if the state is `:accepted`

Via this route this should not be possible to change state from `:accepted` to `:review`.
Via this route it's also not possible for concurrent requests to be racy as this is inside a transaction.

Back to square one I would say. What happened? :-)

**Comment by hennevogel** [comment]
This is what people have done

|time|user|command|
|---|---|---|
08:06:04|licensedigger|osc review accept -G legal-auto 628349
08:06:22|factory-auto| osc review add -G opensuse-review-team 628349
08:06:23|dimstar_suse| osc request accept -f 628349
08:06:26|factory-auto| osc review add -U repo-checker 628349
08:06:28|factory-auto| osc review accept -G factory-auto 628349
08:13:55|namtrac| osc review accept -G opensuse-review-team 628349
09:20:45|staging-bot| osc review add -P openSUSE:Factory:Staging:adi:35 628349
09:20:46|staging-bot| osc review accept -G factory-staging 628349
09:26:04|dimstar| notices something is wrong
```

**Comment by hennevogel** [comment]
I guess the reason is that for adding a review the [permission check runs without a lock](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/bs_request.rb#L823). 

1. For step 2 the [permission check](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/bs_request_permission_check.rb#L219) reads `BsRequest.state` as `:review` and continues to do it's thing
2. While that is running, step 3 kicks in and changes `BsRequest.state` to `:accepted`
3. The permission check is done and `BsRequest.add_review` continues and sets `BsRequest.state` to `:review` again

**Comment by hennevogel** [comment]
I guess the reason is that for adding a review the [permission check runs without a lock](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/bs_request.rb#L823). 

1. For step 2 the [permission check](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/bs_request_permission_check.rb#L219) reads `BsRequest.state` as `:review` and continues to do it's thing
2. While that is running, step 3 kicks in and changes `BsRequest.state` to `:accepted`
3. The permission check is done and `BsRequest.add_review` continues and sets `BsRequest.state` to `:review` again

**Comment by hennevogel** [comment]
`BsRequest.add_review` will even happily wait for step 3, as step 3 runs in a transaction that locks the `BsRequest`. You just have to sneak in step 3 before `BsRequest.add_review` continues to do it's thing after the permission check with a lock.

**Comment by adrianschroeter** [comment]
On Mittwoch, 15. August 2018, 18:00:00 CEST wrote Henne Vogelsang:
> This is what people have done
> 
> |time|user|command|
> |---|---|---|
> 08:06:04|licensedigger|osc review accept -G legal-auto 628349
> 08:06:22|factory-auto| osc review add -G opensuse-review-team 628349
> 08:06:23|dimstar_suse| osc request accept -f 628349
> 08:06:26|factory-auto| osc review add -U repo-checker 628349
> 08:06:28|factory-auto| osc review accept -G factory-auto 628349
> 08:13:55|namtrac| osc review accept -G opensuse-review-team 628349
> 09:20:45|staging-bot| osc review add -P openSUSE:Factory:Staging:adi:35 628349
> 09:20:46|staging-bot| osc review accept -G factory-staging 628349
> 09:26:04|dimstar| notices something is wrong


so it was an addreview call and not a changereviewstate ...

do you see in the log if they did run sequential or parallel?

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by vpereira** [comment]
@hennevogel  that's was exactly what I wrote here: https://github.com/openSUSE/open-build-service/issues/5594#issuecomment-412494154

**Comment by vpereira** [comment]
> I guess the reason is that for adding a review the permission check runs without a lock.
> 
> For step 2 the permission check reads BsRequest.state as :review and continues to do it's thing
> While that is running, step 3 kicks in and changes BsRequest.state to :accepted
> The permission check is done and BsRequest.add_review continues and sets BsRequest.state to :review again

@hennevogel  that's was exactly what I wrote here: https://github.com/openSUSE/open-build-service/issues/5594#issuecomment-412494154

**Comment by vpereira** [comment]
it was fixed with #5655
